include/mfmg/utils.cuh

```diff
@@ -0,0 +1,146 @@
+  cudaError_t error_code = cudaMalloc(&val_dev, nnz * sizeof(ScalarType));
+  ASSERT_CUDA(error_code);
+  error_code = cudaMemcpy(val_dev, &val[0], nnz * sizeof(ScalarType),
+                          cudaMemcpyHostToDevice);
+  ASSERT_CUDA(error_code);
+
+  // Copy the column indices to the gpu
+  int *column_index_dev;
+  error_code = cudaMalloc(&column_index_dev, nnz * sizeof(int));
+  ASSERT_CUDA(error_code);
+  error_code = cudaMemcpy(column_index_dev, &column_index[0], nnz * sizeof(int),
+                          cudaMemcpyHostToDevice);
+  ASSERT_CUDA(error_code);
+
+  // Copy the row pointer to the gpu
+  int *row_ptr_dev;
+  error_code = cudaMalloc(&row_ptr_dev, row_ptr_size * sizeof(int));
+  ASSERT_CUDA(error_code);
+  error_code = cudaMemcpy(row_ptr_dev, &row_ptr[0], row_ptr_size * sizeof(int),
+                          cudaMemcpyHostToDevice);
+  ASSERT_CUDA(error_code);
+
+  return std::make_tuple(val_dev, column_index_dev, row_ptr_dev);
+}
+}
+
+/**
+ * Convert a dealii::SparseMatrix to the regular CSR format and move the data to
+ * the GPU.
+ */
+template <typename ScalarType>
+std::tuple<ScalarType *, int *, int *>
+convert_matrix(dealii::SparseMatrix<ScalarType> const &sparse_matrix)
+{
+  unsigned int const nnz = sparse_matrix.n_nonzero_elements();
+  int const n_rows = sparse_matrix.m();
+  int const row_ptr_size = n_rows + 1;
+  std::vector<ScalarType> val;
+  val.reserve(nnz);
+  std::vector<int> column_index;
+  column_index.reserve(nnz);
+  std::vector<int> row_ptr(row_ptr_size, 0);
+
+  // deal.II stores the diagonal first in each row so we need to do some
+  // reordering
+  for (int row = 0; row < n_rows; ++row)
+  {
+    auto p_end = sparse_matrix.end(row);
+    unsigned int counter = 0;
+    for (auto p = sparse_matrix.begin(row); p != p_end; ++p)
+    {
+      val.emplace_back(p->value());
+      column_index.emplace_back(p->column());
+      ++counter;
+    }
+    row_ptr[row + 1] = row_ptr[row] + counter;
+
+    // Sort the elements in the row
+    unsigned int const offset = row_ptr[row];
+    int const diag_index = column_index[offset];
+    ScalarType diag_elem = sparse_matrix.diag_element(row);
+    unsigned int pos = 1;
+    while ((column_index[offset + pos] < row) && (pos < counter))
+    {
+      val[offset + pos - 1] = val[offset + pos];
+      column_index[offset + pos - 1] = column_index[offset + pos];
+      ++pos;
+    }
+    val[offset + pos - 1] = diag_elem;
+    column_index[offset + pos - 1] = diag_index;
+  }
+
+  return internal::copy_to_gpu(val, column_index, row_ptr);
+}
+
+/**
+ * Move a dealii::TrilinosWrappers::SparseMatrix to the GPU.
+ */
+std::tuple<double *, int *, int *>
+convert_matrix(dealii::TrilinosWrappers::SparseMatrix const &sparse_matrix)
+{
+  unsigned int const n_local_rows = sparse_matrix.local_size();
+  std::vector<double> val;
+  std::vector<int> column_index;
+  std::vector<int> row_ptr(n_local_rows + 1);
+  for (unsigned int row = 0; row < n_local_rows; ++row)
```

Comment:
Isn't there a function to directly get all three arrays out of a Trilinos matrix? There is definitely such a function in Epetra.

Reply:
Why would I want that? I still need to reorder all the column indices.

Reply:
You don't reorder for Trilinos from what I see.

Reply:
Actually `column_index` is not the same as the underlying `Tpetra` array that's why I need to call `global_column_index`. This function just call `sparse_matrix.GCID(indices[i])` ( or `sparse_matrix.GCID64(indices[i])`)